### PR TITLE
Support for match expression

### LINKED ISSUE #24200: [GSOC 2023] Enhance SQLNodeConverterEngine to support more MySQL SQL statements
# Apache ShardingSphere

Apache ShardingSphere is positioned as a Database Plus, and aims at building a standard layer and ecosystem above heterogeneous databases. It focuses on how to reuse existing databases and their respective upper layer, rather than creating a new database. The goal is to minimize or eliminate the challenges caused by underlying databases fragmentation.

Page: https://shardingsphere.apache.org/
Github: https://github.com/apache/shardingsphere 

# Background

The ShardingSphere SQL federation engine provides support for complex SQL statements, and it can well support cross-database join queries, subqueries, aggregation queries and other statements. An important part of SQL federation engine is to convert the SQL statement parsed by ShardingSphere into SqlNode, so that Calcite can be used to implement SQL optimization and federated query.

# Task

This issue is to solve the MySQL exception that occurs during SQLNodeConverterEngine conversion. The specific case list is as follows.

- [x] select_char
- [x] select_extract
- [x] select_from_dual
- [x] select_from_with_table
- [x] select_group_by_with_having_and_window
- [x] select_not_between_with_single_table
- [x] select_not_in_with_single_table
- [x] select_substring
- [x] select_trim
- [x] select_weight_string
- [x] select_where_with_bit_expr_with_ampersand
- [x] select_where_with_bit_expr_with_caret
- [x] select_where_with_bit_expr_with_div
- [x] select_where_with_bit_expr_with_minus_interval
- [x] select_where_with_bit_expr_with_mod
- [x] select_where_with_bit_expr_with_mod_sign
- [x] select_where_with_bit_expr_with_plus_interval
- [x] select_where_with_bit_expr_with_signed_left_shift
- [x] select_where_with_bit_expr_with_signed_right_shift
- [x] select_where_with_bit_expr_with_vertical_bar
- [x] select_where_with_boolean_primary_with_comparison_subquery
- [x] select_where_with_boolean_primary_with_is
- [x] select_where_with_boolean_primary_with_is_not
- [ ] select_where_with_boolean_primary_with_null_safe
- [ ] select_where_with_expr_with_and_sign
- [x] select_where_with_expr_with_is
- [x] select_where_with_expr_with_is_not
- [x] select_where_with_expr_with_not
- [x] select_where_with_expr_with_not_sign
- [x] select_where_with_expr_with_or_sign
- [ ] select_where_with_expr_with_xor
- [x] select_where_with_predicate_with_in_subquery
- [ ] select_where_with_predicate_with_regexp
- [ ] select_where_with_predicate_with_sounds_like
- [ ] select_where_with_simple_expr_with_collate
- [ ] select_where_with_simple_expr_with_match
- [ ] select_where_with_simple_expr_with_not
- [ ] select_where_with_simple_expr_with_odbc_escape_syntax
- [ ] select_where_with_simple_expr_with_row
- [ ] select_where_with_simple_expr_with_tilde
- [ ] select_where_with_simple_expr_with_variable
- [ ] select_window_function
- [ ] select_with_assignment_operator
- [ ] select_with_assignment_operator_and_keyword
- [ ] select_with_case_expression
- [ ] select_with_collate_with_marker
- [ ] select_with_date_format_function
- [ ] select_with_exists_sub_query_with_project
- [ ] select_with_function_name
- [ ] select_with_json_value_return_type
- [ ] select_with_match_against
- [ ] select_with_regexp
- [ ] select_with_schema_name_in_column_projection
- [ ] select_with_schema_name_in_shorthand_projection
- [x] select_with_spatial_function
- [x] select_with_trim_expr
- [x] select_with_trim_expr_from_expr

You need to compare the difference between actual and expected, and then correct the logic in SQLNodeConverterEngine so that actual can be consistent with expected.

After you make changes, remember to add case to `SUPPORTED_SQL_CASE_IDS` to ensure it can be tested.
 
Notice, these issues can be a good example.
- [https://github.com/apache/shardingsphere/pull/14492](https://github.com/apache/shardingsphere/pull/15331)

# Relevant Skills

1. Master JAVA language
2. Have a basic understanding of Antlr g4 file
3. Be familiar with MySQL and Calcite SqlNode

# Targets files

SQLNodeConverterEngineIT - https://github.com/apache/shardingsphere/blob/master/test/it/optimizer/src/test/java/org/apache/shardingsphere/test/it/optimize/SQLNodeConverterEngineIT.java

### PR DESCRIPTION
**Pull request by kanha-gupta**
Ref #24200 .
Support for case ID : select_where_with_simple_expr_with_match

Changes proposed in this pull request:

--Created 4 new files
1. MatchExpression.java for segment support
2. MatchExpressionConverter to get SQLNode
3. MySQLMatchAgainstFunction to add support for the Function
4. ExpectedMatchExpression to add support for XML test case

--Optimised Related files

Before committing this PR, I'm sure that I have checked the following options:
- [x] My code follows the [code of conduct](https://shardingsphere.apache.org/community/en/involved/conduct/code/) of this project.
- [x] I have self-reviewed the commit code.
- [x] I have (or in comment I request) added corresponding labels for the pull request.
- [x] I have passed maven check locally : `./mvnw clean install -B -T1C -Dmaven.javadoc.skip -Dmaven.jacoco.skip -e`.
- [ ] I have made corresponding changes to the documentation.
- [x] I have added corresponding unit tests for my changes.

### CODE REVIEW

**Comment by kanha-gupta** [comment]
@strongduanmu SQLNodeConverterEngineIT tests are successful. 
InternalSQLParserIT is giving error java.lang.UnsupportedOperationException: Unsupported expression: org.apache.shardingsphere.sql.parser.sql.common.segment.dml.expr.MatchExpression
I request you to guide regarding how to add support for new MatchExpression in xml ? 
Thank you :)

**Comment by strongduanmu** [comment]
> @strongduanmu This PR is for code demonstration right now, will optimise it further. SQLNodeConverterEngineIT tests are successful. InternalSQLParserIT is giving error java.lang.UnsupportedOperationException: Unsupported expression: org.apache.shardingsphere.sql.parser.sql.common.segment.dml.expr.MatchExpression I request you to guide regarding how to add support for new MatchExpression in xml ? Thank you :)

Ok, I will take a look at this pr. Thank you for your contribution.

**Comment by kanha-gupta** [comment]
@strongduanmu Please review, Errors are resolved now :)

**Comment by kanha-gupta** [comment]
@strongduanmu Please review, Errors are resolved now :)
PR conclusion - 
1. Adds support for MatchExpression
2.Created a new MySQLMatchAgainst function with reference to calcite
3. Created MatchExpression Converter for conversion to SQLNode
4. Created 2 more files to add support for Testing 
5. Adds MatchExpression support for TestFiles 
6. Optimised XML files

**Comment by kanha-gupta** [comment]
@strongduanmu Please review, Errors are resolved now :)
PR conclusion - 
1. Adds support for MatchExpression
2.Created a new MySQLMatchAgainst function with reference to calcite
3. Created MatchExpression Converter for conversion to SQLNode
4. Created 2 more files to add support for Testing 
5. Adds MatchExpression support for TestFiles 
6. Optimised XML files

**Comment by strongduanmu** [comment]
Hi @kanha-gupta, can you merge master and solve code conflict?

**Comment by kanha-gupta** [comment]
@strongduanmu Please review. Build is successful now :)

File: kernel/sql-federation/core/src/main/java/org/apache/shardingsphere/sqlfederation/compiler/converter/segment/expression/impl/MatchExpressionConverter.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ * Match expression converter.
+ */
+public final class MatchExpressionConverter implements SQLSegmentConverter<MatchExpression, SqlNode> {
+    
+    @Override
+    public Optional<SqlNode> convert(final MatchExpression segment) {
+        List<SqlNode> column = new LinkedList<>();
+        column.add(new SqlIdentifier(segment.getColumnName().getIdentifier().getValue(), SqlParserPos.ZERO));
+        SqlNode against = new ExpressionConverter().convert(segment.getExpr()).get();
+        SqlNode searchModifier = SqlLiteral.createCharString("IN NATURAL LANGUAGE MODE", SqlParserPos.ZERO);
```
Comment (strongduanmu) [inline kernel/sql-federation/core/src/main/java/org/apache/shardingsphere/sqlfederation/compiler/converter/segment/expression/impl/MatchExpressionConverter.java:44]: Why not parse `IN NATURAL LANGUAGE MODE` to MatchExpression?

File: parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/segment/dml/expr/MatchExpression.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    private final int startIndex;
+    
+    private final int stopIndex;
+    
+    private final ColumnSegment columnName;
+    
+    private final ExpressionSegment expr;
+    
+    @Override
+    public String getText() {
```
Comment (strongduanmu) [inline parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/segment/dml/expr/MatchExpression.java:39]: Why return null?

File: kernel/sql-federation/core/src/main/java/org/apache/shardingsphere/sqlfederation/compiler/converter/segment/expression/impl/MySQLMatchAgainstFunction.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+import org.apache.calcite.util.Util;
+
+import java.util.List;
+
+/**
+ * MySQL match against function.
+ */
+public final class MySQLMatchAgainstFunction extends SqlFunction {
+    
+    private static final SqlWriter.FrameType FRAME_TYPE =
```
Comment (strongduanmu) [inline kernel/sql-federation/core/src/main/java/org/apache/shardingsphere/sqlfederation/compiler/converter/segment/expression/impl/MySQLMatchAgainstFunction.java:42]: Please use static import to remove `SqlWriter.`

File: parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/segment/dml/expr/MatchExpression.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+@Setter
+public final class MatchExpression implements ExpressionSegment {
+    
+    private final int startIndex;
+    
+    private final int stopIndex;
+    
+    private final ColumnSegment columnName;
+    
+    private final ExpressionSegment expr;
```
Comment (strongduanmu) [inline parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/segment/dml/expr/MatchExpression.java:36]: Can you refer this class to implement MatchExpression? And then rename this class to MatchAgainstExpression.

https://github.com/polardb/polardbx-sql/blob/main/polardbx-parser/src/main/java/com/alibaba/polardbx/druid/sql/ast/expr/SQLMatchAgainstExpr.java

**Review by strongduanmu — CHANGES_REQUESTED** [review]

**Comment by kanha-gupta** [comment]
@strongduanmu please review, Changes have been made :)

File: kernel/sql-federation/core/src/main/java/org/apache/shardingsphere/sqlfederation/compiler/converter/segment/expression/impl/MySQLMatchAgainstFunction.java (hunk 3 above)
Comment (kanha-gupta) [reply kernel/sql-federation/core/src/main/java/org/apache/shardingsphere/sqlfederation/compiler/converter/segment/expression/impl/MySQLMatchAgainstFunction.java]: done

File: parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/segment/dml/expr/MatchExpression.java (hunk 4 above)
Comment (kanha-gupta) [reply parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/segment/dml/expr/MatchExpression.java]: done

**Review by strongduanmu — APPROVED** [review]
Looks great, merged.

**Pull request merged**